Add HomeWizard Energy integration

DCSBL opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add a new integration for the HomeWizard Energy platform (Wifi P1-Meter, kWh meters and Energy Socket).

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository] > https://github.com/home-assistant/home-assistant.io/pull/20939

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

DCSBL committed:
Start without platform

DCSBL committed:
Update component to make it more resilient

DCSBL committed:
Add unit tests for manual config-flow

DCSBL committed:
Update sensor to comply with 2021.9

DCSBL committed:
Clean up config flow

DCSBL committed:
Update docstrings

DCSBL committed:
Merge PR#56 from DCSBL/ha-homewizard-energy

DCSBL committed:
Fixed picking wrong item for serial

DCSBL committed:
Add config_flow unit test with full coverage

DCSBL committed:
Remove logger from tests

DCSBL committed:
Add unit tests for __init__

DCSBL committed:
Split coordinator into a seperate file

DCSBL committed:
Renamed _host to host

DCSBL commented:
I am not sure if/when to open a PR's to add documentation and move the [brand](https://github.com/home-assistant/brands) from `custom_integration` to `core_integration` (and other PR's if needed). Can I also get some help with that?

DCSBL committed:
Add untested files

DCSBL committed:
Remove switch platform to help with the first review

DCSBL committed:
Remove UNLOAD_CB as ioptions were removed

DCSBL committed:
Add __init__.py to test files

DCSBL committed:
Remove icons for entities with device_class

DCSBL committed:
Remove 'issue_tracker' and update 'documentation' URL

DCSBL commented:
When desired, it is possible to test the integration without having physical hardware. I've created a simple server that simulates the API. -> https://github.com/DCSBL/homewizard-energy-api-sim

DCSBL committed:
Store configuration after first fetch

farmio commented on homeassistant/components/homewizard_energy/sensor.py:
```diff
@@ -0,0 +1,228 @@
+    SensorEntityDescription(
+        key=ATTR_WIFI_STRENGTH,
+        name="Wifi Strength",
+        native_unit_of_measurement=PERCENTAGE,
+        device_class=DEVICE_CLASS_SIGNAL_STRENGTH,
```
percent is an invalid unit for device_class signal_strength
https://developers.home-assistant.io/docs/core/entity/sensor#available-device-classes

DCSBL replied:
What do you think that is the best we can do? As the valid by the API is in percentage. Just remove `device_class`?

farmio replied:
I'd remove device_class and configure a nice icon. 
Maybe it can even be converted 🤔

DCSBL replied:
Agreed, I've removed device_class in 2ad596de7f9b5ff0b647f306b4af11631a388d71. Maybe in the future make a conversion to the correct unit.

farmio commented on homeassistant/components/homewizard_energy/sensor.py:
```diff
@@ -0,0 +1,228 @@
+    energy_api = hass.data[DOMAIN][entry.data["unique_id"]][CONF_API]
+    coordinator = hass.data[DOMAIN][entry.data["unique_id"]][COORDINATOR]
+
+    # Fetch initial data so we have data when entities subscribe
+    await coordinator.async_refresh()
```
this should already have happened at
https://github.com/home-assistant/core/blob/a49d707e3042e3862025bc5fdea7831e7fa0c676/homeassistant/components/homewizard_energy/__init__.py#L72

DCSBL replied:
Thanks, fixed in b945a09e690a5c995742befbb4806d31da51a6ca

farmio reviewed (commented):
👍

DCSBL committed:
Apply suggestions from code review

Co-authored-by: Matthias Alphart <farmio@alphart.net>

DCSBL committed:
Update homeassistant/components/homewizard_energy/__init__.py

Co-authored-by: Matthias Alphart <farmio@alphart.net>

DCSBL committed:
Remove CONFIG_SCHEMA

DCSBL committed:
Use f-strings for format

DCSBL committed:
Remove second 'initial' coordinator refresh

DCSBL commented on homeassistant/components/homewizard_energy/config_flow.py:
```diff
@@ -0,0 +1,190 @@
+        # Format name
+        self.context["custom_name"] = user_input["name"]
+        if Lower(self.context["product_name"]) != Lower(user_input["name"]):
+            title = "{} ({})".format(
```
Ah missed that one. Fixed in fd9db6fdc0274a2d501165cb796c92e0263753c5

DCSBL committed:
Remove async_setup

DCSBL commented on homeassistant/components/homewizard_energy/__init__.py:
```diff
@@ -0,0 +1,107 @@
+    _LOGGER.debug("__init__ async_setup")
+    hass.data[DOMAIN] = {}
+
+    return True
```
Didn't know that. Fixed in ae4fc6495345d204448e803ab6f0ba5076bf2f19 and ae4fc6495345d204448e803ab6f0ba5076bf2f19

DCSBL commented:
> 👍

Thank you for your time. I have processed your feedback.

DCSBL committed:
Remove device_class for wifi strength

klaasnicolaas commented:
Just remove it from the branch 😉

DCSBL committed:
Remove config folder

mbbs1024 commented:
From a user point of view, it would be interesting if Home assistant would be able to have access to the full P1 meter parameter list, instead of only the limited list this custom component currently offers.
The full parameters are accessible by using the P1-meter API, that delivers the complete telegram, containing all measurements and data the digital meter is capable of.
http://{IP address}/api/v1/telegram

DCSBL commented:
I would love that, but it would mean that this component needs to parse data from the P1 differently than the other Energy devices (JSON vs raw telegram) which is a bit of a hassle. The support for all entities exposed by JSON is currently in the pipeline by HomeWizard so I think it is not worth it to use the raw telegram data.

klaasnicolaas commented:
This is not the place to talk about features, you can go to the [community forum](https://community.home-assistant.io/) for that. Please stay ontopic in this pull request and that is reviewing the python code 😉

MansM commented:
any chance to get this in before the 2021.12 release. So SintaMans or SantaMans can be nice to me ;-)

DCSBL committed:
Select better icons for sensors

DCSBL committed:
Add entity categories

DCSBL committed:
Remove zeroconf dependency

DCSBL committed:
Add tests for all files

DCSBL committed:
Remove duplicate requirements

DCSBL commented:
```
----------- coverage: platform linux, python 3.9.7-final-0 -----------
Name                                                        Stmts   Miss  Cover   Missing
-----------------------------------------------------------------------------------------
homeassistant/components/homewizard_energy/__init__.py         48      0   100%
homeassistant/components/homewizard_energy/config_flow.py      80      0   100%
homeassistant/components/homewizard_energy/const.py            32      0   100%
homeassistant/components/homewizard_energy/coordinator.py      42      0   100%
homeassistant/components/homewizard_energy/sensor.py           50      0   100%
-----------------------------------------------------------------------------------------
TOTAL                                                         252      0   100%


Results (5.59s):
      35 passed
```

DCSBL committed:
Fix test issues

balloob commented on CODEOWNERS:
```diff
@@ -228,6 +228,7 @@ homeassistant/components/homeassistant/* @home-assistant/core
 homeassistant/components/homekit/* @bdraco
 homeassistant/components/homekit_controller/* @Jc2k @bdraco
 homeassistant/components/homematic/* @pvizeli @danielperna84
+homeassistant/components/homewizard_energy/* @DCSBL
```
This integration should be called `homewizard` as that's the brand name.

DCSBL replied:
HomeWizard is the company and it also has an (obsolete) product called 'HomeWizard'. This integration is developed for a new product category 'HomeWizard Energy' and is not developed with the intention to support other products than the products with this name.

IMO, it is confusing to have the name `homewizard` as it implies that is designed for the old 'HomeWizard'. And 'energy` is to generic (and also used of course). What do you think?

balloob replied:
Let's stick with the company name. We can just document it's for energy devices only.

DCSBL replied:
Changed domain to `homewizard` (https://github.com/home-assistant/core/pull/55812/commits/54648e39477377658d8a659b8de6b8e80ca5dad3). I think we should update some metadata for PR now, what is the best approach?
 
The brand should be added/changed to. Do we stick with the HomeWizard Energy logo (Like 'the Philips Hue' logo) or do we use the HomeWizard logo (like a 'Philips' logo)

balloob commented on homeassistant/components/homewizard_energy/config_flow.py:
```diff
@@ -0,0 +1,187 @@
+        entry_info["product_type"] = energy_api.device.product_type
+        entry_info["serial"] = energy_api.device.serial
+
+        self.context[CONF_HOST] = entry_info[CONF_IP_ADDRESS]
```
Don't store this in context but just on this config flow instance

DCSBL replied:
Removed it with 79f574f2d05a226bddeb6ec3016f667565132e72 and https://github.com/home-assistant/core/pull/55812/commits/edac621182d6f9cbf1d45c389e453d479395844f

balloob commented on homeassistant/components/homewizard_energy/__init__.py:
```diff
@@ -0,0 +1,107 @@
+    await coordinator.async_config_entry_first_refresh()
+
+    # Finalize
+    hass.data[DOMAIN][entry.data["unique_id"]] = {
```
You can also use `entry.entry_id` as a unique ID.

balloob commented on homeassistant/components/homewizard_energy/config_flow.py:
```diff
@@ -0,0 +1,187 @@
+                description_placeholders={"name": self.context["product_name"]},
+                data_schema=Schema(
+                    {
+                        Required("name", default=self.context["product_name"]): All(
```
No need to ask for the name. Entries can be renamed in the UI later.

DCSBL replied:
I implemented this to have a useful entity name when you add more that one energy device. Otherwise you have to rename all the entities to something like `sensor.refrigerator_active_power` etc.

Quitte a few users that use the custom integration told me that they like to have the name to be added to help with the problem above. It is there a better way to do that?

It is currently not possible to get the name the user had entered from the app, so we can't use that at this moment.

balloob replied:
When renaming a device in the UI the entities are renamed too.

DCSBL replied:
Cool, didn't know that.

DCSBL replied:
Removed custom name stuff with https://github.com/home-assistant/core/pull/55812/commits/79f574f2d05a226bddeb6ec3016f667565132e72

balloob commented on homeassistant/components/homewizard_energy/config_flow.py:
```diff
@@ -0,0 +1,187 @@
+            CONF_PORT: 80,
+        }
+
+        return await self.async_step_check(entry_info)
```
No need to ask for confirmation. If the entered IP address works, add it.

DCSBL replied:
It now asks for the custom name and then adds it. It won't ask you to confirmation. Maybe the names are wrong.

This can be changed based on what you decide with https://github.com/home-assistant/core/pull/55812#discussion_r762157217

DCSBL replied:
Removed confirmation and custom name with https://github.com/home-assistant/core/pull/55812/commits/79f574f2d05a226bddeb6ec3016f667565132e72

balloob commented on homeassistant/components/homewizard_energy/coordinator.py:
```diff
@@ -0,0 +1,93 @@
+            try:
+                smr_version = self.api.data.smr_version
+                if smr_version == 50:
+                    return timedelta(seconds=1)
```
We do not allow polling API update intervals lower than 5 seconds.

DCSBL replied:
Fixed by [14897528a](https://github.com/DCSBL/core/commit/14897528ab2c4441b69f21177c505372a2fc2796) and [9cf9c0e7a](https://github.com/DCSBL/core/commit/9cf9c0e7a26000af2cd4d132c98a1ee81044374a)

balloob commented on homeassistant/components/homewizard_energy/coordinator.py:
```diff
@@ -0,0 +1,93 @@
+                status = await self.api.update()
+
+                if not status:
+                    raise Exception("Failed to fetch data")
```
Raise `UpdateFailed` directly and drop the try…except

DCSBL replied:
Changed with [d7a09d6c](https://github.com/DCSBL/core/commit/d7a09d6c5feb8f31589aa2e2294418fd3dbcc22d). But I think we still need to catch the [timeout](https://github.com/DCSBL/core/blob/d7a09d6c5feb8f31589aa2e2294418fd3dbcc22d/homeassistant/components/homewizard_energy/coordinator.py#L47)?

balloob replied:
[Nope](https://github.com/home-assistant/core/blob/dev/homeassistant/helpers/update_coordinator.py#L187-L189)

DCSBL replied:
O nice! Removed the try..catch 👍 https://github.com/home-assistant/core/pull/55812/commits/1688b2826f8b2a1d2feef52a9d49f4aa23026986

balloob commented on homeassistant/components/homewizard_energy/coordinator.py:
```diff
@@ -0,0 +1,93 @@
+                    raise Exception("Failed to fetch data")
+
+                data = {
+                    CONF_NAME: self.api.device.product_name,
```
Don't include static information in the returned data. Instead, let entities that need this data access it on the coordinator. You can add helper properties on the coordinator if it helps clean up the code.

DCSBL replied:
Agree with https://github.com/home-assistant/core/pull/55812/commits/c289f66a573b1e7de465d5c68f9a54cdcc90576f ?

balloob commented on homeassistant/components/homewizard_energy/sensor.py:
```diff
@@ -0,0 +1,226 @@
+        entity_category=ENTITY_CATEGORY_DIAGNOSTIC,
+    ),
+    SensorEntityDescription(
+        key=ATTR_WIFI_STRENGTH,
```
Disable this one by default. It spams.

DCSBL replied:
https://github.com/DCSBL/core/commit/2b74b7715dd2e2d630a1f2bba1d7fdd24c195c7f

balloob commented on homeassistant/components/homewizard_energy/sensor.py:
```diff
@@ -0,0 +1,226 @@
+        if description.key in energy_api.data.available_datapoints:
+            entities.append(HWEnergySensor(coordinator, entry.data, description))
+    async_add_entities(entities, update_before_add=True)
+
+    return True
```
No return value on async_setup_entry for platforms.
```suggestion
```

balloob commented on homeassistant/components/homewizard_energy/sensor.py:
```diff
@@ -0,0 +1,226 @@
+    for description in SENSORS:
+        if description.key in energy_api.data.available_datapoints:
+            entities.append(HWEnergySensor(coordinator, entry.data, description))
+    async_add_entities(entities, update_before_add=True)
```
Why is `update_before_add` set to True? You're using an update coordinator and all data has been fetched during startup

DCSBL replied:
I think I've added this before I added the coordinator. Removed with https://github.com/DCSBL/core/commit/d98cba8f4f7af84caefbb39457eb5900ac6eab72

balloob commented on homeassistant/components/homewizard_energy/sensor.py:
```diff
@@ -0,0 +1,226 @@
+
+        super().__init__(coordinator)
+        self.entity_description = description
+        self.coordinator = coordinator
```
Not necessary, `CoordinatorEntity` already does this.

DCSBL replied:
Thanks. Removed: https://github.com/DCSBL/core/commit/fbf6622200456307b095ad1d888add7c22955f30

balloob commented on homeassistant/components/homewizard_energy/sensor.py:
```diff
@@ -0,0 +1,226 @@
+        """Return data object from DataUpdateCoordinator."""
+        return self.coordinator.data
+
+    @property
+    def icon(self):
+        """Return the icon."""
+        return self.entity_description.icon
+
```
This the default behavior
```suggestion
```

balloob commented on homeassistant/components/homewizard_energy/sensor.py:
```diff
@@ -0,0 +1,226 @@
+        self.entry_data = entry_data
+
+        # Config attributes.
+        self.name = "{} {}".format(entry_data["custom_name"], description.name)
```
Assign to `self._attr_name` to have the name property pick it up. 

Custom name should be dropped, just use the config entry title.

DCSBL replied:
Dropped custom name with https://github.com/home-assistant/core/pull/55812/commits/79f574f2d05a226bddeb6ec3016f667565132e72. But I can't get the name to work when I use `self._attr_name`. It just uses an 'readable' unique id.

<img width="1060" alt="Screenshot 2021-12-06 at 21 16 39" src="https://user-images.githubusercontent.com/74970928/144916013-9eba0fc5-0530-40b2-a40a-1e314a7a335f.png">
.

DCSBL replied:
https://github.com/home-assistant/core/pull/55812/commits/6fba568fdf44bd1c329f30b57099aae373a25ef4

balloob commented on homeassistant/components/homewizard_energy/strings.json:
```diff
@@ -0,0 +1,14 @@
+      }
+    },
+    "abort": {
+      "single_instance_allowed": "[%key:common::config_flow::abort::single_instance_allowed%]",
+      "no_devices_found": "[%key:common::config_flow::abort::no_devices_found%]"
```
these are not used. already_configured is used.

DCSBL replied:
Added used translations and removed unused translations with https://github.com/DCSBL/core/commit/13b652b581b45f4cf15cd4c06704f2af0f2b8359

balloob commented on homeassistant/components/homewizard_energy/strings.json:
```diff
@@ -0,0 +1,14 @@
+{
+  "title": "Homewizard Energy",
+  "config": {
+    "step": {
```
ip is missing.

balloob commented on homeassistant/components/homewizard_energy/translations/en.json:
```diff
@@ -0,0 +1,27 @@
+{
```
This file will be automatically generated based on `strings.json`. During development you can use `python3 -m script.translations.develop` to re-generate this file.

DCSBL replied:
Thanks. If I understand it correctly, I don't have to add the `translations/*.json` files to this PR because they are generated by the strings.json file?

balloob replied:
Correct.

DCSBL replied:
Updated `strings.json` and removed `translations/*.json` https://github.com/home-assistant/core/pull/55812/commits/ca98ab80a2112a767e03b825612eabfbd1251d2c

balloob commented on homeassistant/components/homewizard_energy/translations/nl.json:
```diff
@@ -0,0 +1,29 @@
+{
+    "title": "Homewizard Energy",
```
These translations will be overridden by the translations in lokalise.

balloob commented:
I see that you're affiliated with HomeWizard. is this an official integration?

DCSBL commented:
@balloob thank you for the review, I think I can process it coming week.

> I see that you're affiliated with HomeWizard. is this an official integration?

This is not an official integration. I made this in my spare time. But maybe I can see if we can make it ‘official’ if that is desirable.

DCSBL committed:
Set minimum update interval to 5 seconds

DCSBL committed:
Call UpdateFailed directly

DCSBL committed:
Disable wifi strength by default

DCSBL committed:
Remove return value from async_setup_entry()

Co-authored-by: Paulus Schoutsen <paulus@home-assistant.io>

DCSBL committed:
Simplify update_interval

DCSBL committed:
Remove update_before_add

DCSBL committed:
Remove useless assignment, already handled by CoordinatoEntity

DCSBL committed:
Update homeassistant/components/homewizard_energy/sensor.py

Co-authored-by: Paulus Schoutsen <paulus@home-assistant.io>

DCSBL committed:
Attempt to fix translations

DCSBL committed:
Remove try..except for timeout as it is already handled

DCSBL committed:
Refactored config_flow to better handle zeroconf

DCSBL committed:
Use const for common keywords

DCSBL committed:
Fix warning about using def state()

DCSBL committed:
Simplify coordinator

DCSBL committed:
Update tests

DCSBL committed:
Do not set stuff in context

balloob commented on homeassistant/components/homewizard_energy/__init__.py:
```diff
@@ -0,0 +1,98 @@
+
+    except aiohwenergy.DisabledError as ex:
+        _LOGGER.error("API disabled, API must be enabled in the app")
+        raise ConfigEntryNotReady from ex
```
Return `False` because the config entry will not be able to set up and we should stop retrying. Users can reload the entry if they have re-enabled the API.
```suggestion
        return False
```

DCSBL replied:
I try to implement this in while moving the setup to coordinator (requested in https://github.com/home-assistant/core/pull/55812#discussion_r763643640). How can I tell coordinator/HA to stop retrying when it detected a disabled API?

balloob replied:
raise a specific exception that you catch when you call first_refresh

DCSBL replied:
The coordinator raises an `ConfigEntryNotReady` when something when wrong https://github.com/home-assistant/core/blob/af603d04272ec489d03b26cb6384d062611d0a75/homeassistant/helpers/update_coordinator.py#L159-L161
I couldn't figure out how to catch the original exception and can't find a component that tries to. To get the desired behaviour I tried it like this; https://github.com/home-assistant/core/pull/55812/commits/e06bb269948a1073d6bea8f3b240e6c8587aab29. What do you think?

balloob commented on homeassistant/components/homewizard_energy/__init__.py:
```diff
@@ -0,0 +1,98 @@
+    initialized = False
+    try:
+        with async_timeout.timeout(10):
+            await energy_api.initialize()
```
I would move this into the coordinator, and then you can let `await coordinator.async_config_entry_first_refresh()` do it's magic. Create `energy_api` object inside the coordinator update method and initialize when created.

DCSBL replied:
Moved init to coordinator (https://github.com/home-assistant/core/pull/55812/commits/3e3fb381b8bf9472f84a236ac6e5557694611407)

balloob commented on homeassistant/components/homewizard_energy/config_flow.py:
```diff
@@ -0,0 +1,182 @@
+            CONF_PORT: 80,
+            CONF_PRODUCT_TYPE: device_info[CONF_PRODUCT_TYPE],
+            CONF_PRODUCT_NAME: device_info[CONF_PRODUCT_NAME],
+            CONF_SERIAL: device_info[CONF_SERIAL],
```
Set this as unique ID.

DCSBL replied:
https://github.com/home-assistant/core/pull/55812/commits/c7c7ceb415e7e37ce43bdaf92d6a271fe484f177

balloob commented on homeassistant/components/homewizard_energy/__init__.py:
```diff
@@ -0,0 +1,98 @@
+
+    # Finalize
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][entry.unique_id] = {
```
Why not use `entry.entry_id` ?

DCSBL replied:
https://github.com/home-assistant/core/pull/55812/commits/6b5b808f4d145182c7815e3878631496f556ec15

balloob commented on homeassistant/components/homewizard_energy/config_flow.py:
```diff
@@ -0,0 +1,182 @@
+            or "path" not in discovery_info["properties"]
+            or "product_name" not in discovery_info["properties"]
+            or "product_type" not in discovery_info["properties"]
+            or "serial" not in discovery_info["properties"]
```
You alraedy have serial here, set it and let it abort right away if it's already configured to avoid having to try to connect.

DCSBL replied:
https://github.com/home-assistant/core/pull/55812/commits/c7c7ceb415e7e37ce43bdaf92d6a271fe484f177

balloob commented on homeassistant/components/homewizard_energy/config_flow.py:
```diff
@@ -0,0 +1,182 @@
+        return {
+            CONF_PRODUCT_NAME: energy_api.device.product_name,
+            CONF_PRODUCT_TYPE: energy_api.device.product_type,
+            CONF_SERIAL: energy_api.device.serial,
```
Why store the serial in the config entry? You're going to fetch it each time in the coordinator and it's also stored as unique ID.

DCSBL replied:
https://github.com/home-assistant/core/pull/55812/commits/fc1c4b7c74592b33f940f24046040973fc9cefed

balloob commented on homeassistant/components/homewizard_energy/sensor.py:
```diff
@@ -0,0 +1,217 @@
+        state_class=STATE_CLASS_MEASUREMENT,
+    ),
+    SensorEntityDescription(
+        key=ATTR_ACTIVE_POWER_L1_W,
```
I would just inline these constants because these classes itself are constants

DCSBL replied:
You mean replace things like `ATTR_ACTIVE_POWER_L1_W` to `"active_power_l1_w"`?

balloob replied:
Personal preference. These objects are constants and the name is also here. It's not used anywhere else. Now the definition of these objects is unnecessarily spread over 2 files.

DCSBL replied:
Just to be sure, I understand that this (-> https://github.com/home-assistant/core/pull/55812/commits/f297592bbc2d11e17ec7830f49cfaf94231cabe8) is what you prefer?

DCSBL replied:
👍

balloob commented on homeassistant/components/homewizard_energy/sensor.py:
```diff
@@ -0,0 +1,217 @@
+class HWEnergySensor(CoordinatorEntity, SensorEntity):
+    """Representation of a HomeWizard Energy Sensor."""
+
+    unique_id = None
+    name = None
```
Use the existing `_attr_unique_id` and `_attr_name` equivalents.

DCSBL replied:
https://github.com/home-assistant/core/pull/55812/commits/6fba568fdf44bd1c329f30b57099aae373a25ef4

balloob commented on homeassistant/components/homewizard_energy/sensor.py:
```diff
@@ -0,0 +1,217 @@
+        self.entry_data = entry.data
+
+        # Config attributes.
+        self.name = "{} {}".format(entry.data["name"], description.name)
```
Use string formatting.
```suggestion
        self._attr_name = f"{entry.data['name']} {description.name}"
```

DCSBL committed:
Move initial setup in coordinator

DCSBL committed:
Update homeassistant/components/homewizard_energy/sensor.py

Co-authored-by: Paulus Schoutsen <paulus@home-assistant.io>

DCSBL committed:
Use existing attributes

DCSBL committed:
Check for unique_id in discovery flow before connect

DCSBL committed:
Fix tests

DCSBL committed:
Store only important data as entry

DCSBL committed:
Use entry_id instead of unique_id

DCSBL committed:
Bring back try..except to catch api_disabled exception

DCSBL committed:
Remove translations

DCSBL committed:
Better handle closing connection

DCSBL committed:
Fix pylint issues

DCSBL committed:
Test timestamp for gas_timestamp

DCSBL committed:
Rename 'homewizard_energy' to 'homewizard'

DCSBL committed:
Fix tests for coordinator

DCSBL committed:
Rename all HomeWizard Energy strings to HomeWizard

DCSBL committed:
Set name to coordinator

DCSBL committed:
Don't retry initialization when API is disabled

DCSBL committed:
Inline api data-key names

DCSBL committed:
Remove unused values

DCSBL committed:
Test with API

DCSBL committed:
Fix removed const

DCSBL commented:
The domain is renamed from `homewizard_energy` to `homewizard`, see https://github.com/home-assistant/core/pull/55812#discussion_r762153704.

I think the metadata be updated, what is the best approach?
The brand `homewizard` should be added/changed too. Do we stick with the HomeWizard Energy logo (Like 'the Philips Hue' logo) or do we use the HomeWizard logo (like a 'Philips' logo)?

farmio commented on homeassistant/components/homewizard/coordinator.py:
```diff
@@ -28,15 +28,18 @@ def __init__(
-                self.api = await self._initialize_api()
+                await self.initialize_api()
+
+            # Tell MyPi that self.api is set (silence attr-defined)
+            assert self.api is not None
```
You can retrun the api object from `self.initialize_api()` so Mypy would know without having to use `assert` (which you shouldn't do in production code).

Also please either fully type functions or classes or don't annotate at all (eg. `__init__()` is missing an annotation for `host`) and try to specify generic types like `dict` (in this case eg. `_async_update_data(self) -> dict[str, Any]`)

You may also don't have to use an optional for self.api and self.host - just annotate it at class level `api: aiohwenergy.HomeWizardEnergy` or `host: str` without setting it to None and make sure you don't access it before you assigned something - eg. in `__init__()`. On the other hand - if you assign in `__init__` you can just omit the class level assignment.

DCSBL replied:
Sorry for the delay, but thanks!
I've fixed some issues I had with initialising the API while having the required behaviour. I think it should be fixed now.

I also annotated everything I can think of and passed `mypy`, so I've added it to `.strict-typing`.

DCSBL committed:
Update aiohwenergy to 0.5.1

DCSBL committed:
Update coordinator to not require assert and add type annotations

DCSBL committed:
Add type annotations and fix related issues

DCSBL committed:
Fix failing tests

DCSBL commented:
Changed a few things here and there to better handle initialisation and added type annotations (and fixed issues that came with it). But it ended up to be few a large commits because I it didn't pass the pre-commit steps. Sorry about that.

DCSBL committed:
Bumb aiohwenergy to 0.6.0 (fixes pytest issue on 3.8)

DCSBL committed:
Fix missing pypi version

balloob commented on homeassistant/components/homewizard/sensor.py:
```diff
@@ -0,0 +1,208 @@
+        entity_category=ENTITY_CATEGORY_DIAGNOSTIC,
+    ),
+    SensorEntityDescription(
+        key="meter_model",
```
Why is this a sensor? It is already part of device info and a user will never automate on this.

DCSBL replied:
I think it is better to rename this: This is not the meter model as in "It is an HomeWizard P1 meter", but the smart meter's model number. I feel it is the same as "WiFi SSID" for a user.

<img width="326" alt="Screenshot 2021-12-30 at 20 52 08" src="https://user-images.githubusercontent.com/74970928/147783897-2f6e163c-8330-4120-b73d-23db6cffcadd.png">

I have renamed `meter_model` and `gas_timestamp` entities to make it more clear, what do you think? -> b926571

balloob commented on homeassistant/components/homewizard/sensor.py:
```diff
@@ -0,0 +1,208 @@
+    ),
+    SensorEntityDescription(
+        key="gas_timestamp",
+        name="Gas Timestamp",
```
What does this timestamp represent? This name is not descriptive enough

DCSBL replied:
Agreed with "Last gas update"?
Also set it as it a 'diagnostic' as it does not really represent a measurement. It can be derived from an actual update seen in the 'total_gas' sensor.

Or maybe remove this one at all as all the information is already provided by HA?

balloob commented on homeassistant/components/homewizard/sensor.py:
```diff
@@ -0,0 +1,208 @@
+
+        # Some values are given, but set to NULL (eg. gas_timestamp when no gas meter is connected)
+        if self.data["data"][self.data_type] is None:
+            self.entity_description.entity_registry_enabled_default = False
```
the entity descriptions are part of constants. If a user would set up 2 Homewizard devices, it is now overriden. Make a copy instead.

balloob replied:
Although a better idea would be to not add these entities to begin with ? They will never have data.

DCSBL replied:
Changed it that `None` values won't be added (c593306). The `power_export` enabled_default are now set into `_attr_entity_registry_enabled_default` (a1be964)

balloob reviewed (approved):
Looks good! Some tiny comments left.

DCSBL committed:
Rename sensors to better represent what they do

DCSBL committed:
Don't add entities if value is None

DCSBL committed:
set _attr instead of const for enabled by default

balloob commented on homeassistant/components/homewizard/sensor.py:
```diff
@@ -126,8 +126,9 @@
     SensorEntityDescription(
         key="gas_timestamp",
-        name="Gas Timestamp",
+        name="Last Gas Update",
```
Why do we need this? Isn't this already visible because that's the time the gas usage state got updated in Home Assistant?

DCSBL replied:
I agree. Removed the sensor (f072a7d)

DCSBL committed:
Remove gas_timestamp sensor

DCSBL commented:
@balloob, What is the next step?

balloob commented:
Me actually circling back 😬 

Looks good!

MartinHjelmare reviewed (commented):
Please address the comments in a new PR. Thanks!
